### PR TITLE
Implement server side packet handling for eth/62 and eth/63

### PR DESCRIPTION
**Pull request by karalabe**
This PR aims to implement the "server" side of the [eth/62](https://github.com/ethereum/wiki/wiki/Ethereum-Wire-Protocol#proposed-messages-for-new-model-syncing-pv62) and [eth/63](https://github.com/ethereum/wiki/wiki/Ethereum-Wire-Protocol#proposed-messages-for-light-client-pv63) protocols. The goal of v62 is to introduce the necessary network packets for more flexible full node synchronization, whereas v63 aims to set the bases for implementing light clients. All the proposed network packets have been implemented and handled, also supporting switching the between any of the 4 currently supported protocol versions: 60 through 63.

Note, the client side of the protocols (i.e. synchronization, downloader, fetcher, etc), have **not** yet been updated, so enabling protocols above 62 and connecting to the main net would likely result in very undefined behavior.

However, given the scope of this PR, I would like to have a round of reviews and corrections in the current state, and only move forward with implementing the client side and updated/new sync logic after the server side core is deemed to be solid. That should desirable be done in a new PR to keep the merges clean.

### CODE REVIEW

**Comment by obscuren** [comment]
Squash commits please

File: eth/handler.go [hunk 1]
```diff
@@ -305,51 +302,33 @@ func (pm *ProtocolManager) handleMsg(p *peer) error {
 		var (
 			hash   common.Hash
 			bytes  common.StorageSize
-			hashes []common.Hash
 			blocks []*types.Block
 		)
-		for {
+		for len(blocks) < downloader.MaxBlockFetch && bytes < softResponseLimit {
```
Comment (Gustav-Simonsson) [inline eth/handler.go]: MaxBlockFetch sounds like a inclusive bound, if it's not the comment where it's defined needs an update.

File: eth/handler.go [hunk 2]
```diff
@@ -357,42 +336,201 @@ func (pm *ProtocolManager) handleMsg(p *peer) error {
...
+		if err := msg.Decode(&query); err != nil {
+			return errResp(ErrDecode, "%v: %v", msg, err)
+		}
+		// Gather blocks until the fetch or network limits is reached
+		var (
+			bytes   common.StorageSize
+			headers []*types.Header
+			unknown bool
+		)
+		for !unknown && len(headers) < int(query.Amount) && bytes < softResponseLimit && len(headers) < downloader.MaxHeaderFetch {
```
Comment (Gustav-Simonsson) [inline eth/handler.go]: Same here, if MaxHeaderFetch is not inclusive the comment where it's defined needs an update.

File: eth/handler.go (hunk 1 above)
Comment (karalabe) [inline eth/handler.go]: It is inclusive, isn't it? As long as the length < Max, add some more.

File: eth/handler.go [hunk 3]
```diff
@@ -357,42 +336,201 @@ func (pm *ProtocolManager) handleMsg(p *peer) error {
...
+			// Retrieve the next block satisfying the query
+			var origin *types.Block
+			if query.Origin.Hash != (common.Hash{}) {
```
Comment (Gustav-Simonsson) [inline eth/handler.go]: Parenthesises around common.Hash{} can be removed?

Comment (karalabe) [inline eth/handler.go]: No, because Go would interpret it as the body of the "if".

File: eth/handler.go (hunk 2 above)
Comment (karalabe) [inline eth/handler.go]: Same as above, check my comment ;)

File: eth/handler.go (hunk 3 above)
Comment (obscuren) [inline eth/handler.go]: No they can not. Without that, it will give a syntax error (think about it! ;-))

File: eth/handler.go [hunk 4]
```diff
@@ -357,42 +336,201 @@ func (pm *ProtocolManager) handleMsg(p *peer) error {
...
+			} else {
+				origin = pm.chainman.GetBlockByNumber(query.Origin.Number)
+			}
+			if origin == nil {
+				break
+			}
+			headers = append(headers, origin.Header())
+			bytes += origin.Size()
+
+			// Advance to the next block of the query
```
Comment (Gustav-Simonsson) [inline eth/handler.go]: Why is both hash and block number supported for block traversal?

Comment (karalabe) [inline eth/handler.go]: You need hash to be able to address a fork (i.e. not the main chain), also as light clients don't know/care about the height of the chain, they only see the latest hash propagated and have to be able and fetch stuff (headers) based on it.

Other than those, my algos only use number based retrievals. I think the idea was to make it flexible enough to permit implementers to experiment with different sync algos and not force one specific one on them.

File: eth/handler.go (hunk 2 above)
Comment (Gustav-Simonsson) [inline eth/handler.go]: Ah, you where too quick to respond before I deleted it :) Yepp saw the tests, nevermind!

File: eth/handler.go (hunk 3 above)
Comment (Gustav-Simonsson) [inline eth/handler.go]: Ah, makes sense. Thanks!

File: eth/handler.go (hunk 4 above)
Comment (Gustav-Simonsson) [inline eth/handler.go]: If hash traversal is needed to address a fork, how come block number traversal is supported here? In what cases is block number traversal used?

Comment (karalabe) [inline eth/handler.go]: The downloader always uses numbers to traverse the chain. Number are always based off the canonical chain, so they should be valid up to the very tip of the chain (last few blocks may end up being a side fork). The fetcher will sort out the top blocks after the downloader pulled in the masses.

Comment (Gustav-Simonsson) [inline eth/handler.go]: Interesting. I still don't understand why both approaches are in place, I will have to peruse the downloader and fetcher code more :)

Comment (karalabe) [inline eth/handler.go]: It's the [eth/62](https://github.com/ethereum/wiki/wiki/Ethereum-Wire-Protocol#proposed-messages-for-new-model-syncing-pv62) protocol spec. As I mentioned, I don't use them for anything, but I do need to make sure the spec is honored :)

**Comment by karalabe** [comment]
As the full eth/62 sync logic is already finished in the top level PR, I'm closing this one down in favor of a single big merge now: https://github.com/ethereum/go-ethereum/pull/1701.

**Pull request closed without merging**